### PR TITLE
Code cleanup and add new critical head block and move critical css before assets

### LINKED ISSUE #38748: [Issue] Code cleanup and add new critical head block and move critical css before assets
This issue is automatically created based on existing pull request: magento/magento2#35580: Code cleanup and add new critical head block and move critical css before assets

---------

### Description (*)
The critical css should appear before the assets. In addition, the changes provide many more opportunities to insert custom blocks between the meta details and assets via template or xml blocks. These include, for example, own meta specifications which are not controllable via XML, as well as assets from the template which must be inserted before other files, for example, constentscripts or special preloads, og:meta tags.

These changes are important to make the frontend basically more customizable and performace or function optimization. 

### Fixed Issues (if relevant)

1. use short syntax PHP 5 compatibility is no longer needed
2. adjustment namespace import
3. use switch statement with direct return
4. move critical inline css before assets

### Manual testing scenarios (*)

1. Use CSS critical path

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

Before:
before

After:
after

### PR DESCRIPTION
**Pull request by in-session**
### Description (*)
The critical css should appear before the assets. In addition, the changes provide many more opportunities to insert custom blocks between the meta details and assets via template or xml blocks. These include, for example, own meta specifications which are not controllable via XML, as well as assets from the template which must be inserted before other files, for example, constentscripts or special preloads, og:meta tags.

These changes are important to make the frontend basically more customizable and performace or function optimization. 

### Fixed Issues (if relevant)

1. use short syntax PHP 5 compatibility is no longer needed
2. adjustment namespace import
3. use switch statement with direct return
4. move critical inline css before assets

### Manual testing scenarios (*)

1. Use CSS critical path

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

Before:
before

After:
after

### Resolved issues:
1. [x] resolves magento/magento2#38748: Code cleanup and add new critical head block and move critical css before assets

### CODE REVIEW

**Comment by in-session** [comment]
@magento give me test instance

**Comment by in-session** [comment]
@magento give me test instance

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento give me test instance

**Comment by in-session** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
Hi Just found this update seem quite useful. Have you done from your side and read for test ? @in-session 
Thanks

File: app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml [hunk 1]
```diff
@@ -12,7 +12,7 @@
         <script src="requirejs/require.js"/>
     </head>
     <body>
```
Comment (mrtuvn) [inline app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml:14]: just think introduced/replaced new block can affect to another old code 3rd-party rely on block head.additional

Comment (mrtuvn) [reply app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml:14]: or i missed understood

**Comment by in-session** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Can you check code standards at files changes related ?

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run all tests

**Comment by in-session** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by in-session** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by in-session** [comment]
@magento give me test instance

File: app/code/Magento/Catalog/view/frontend/layout/default.xml [hunk 2]
```diff
@@ -65,6 +65,7 @@
         <referenceContainer name="after.body.start">
             <block class="Magento\Framework\View\Element\Js\Components" name="head.components" as="components" template="Magento_Catalog::js/components.phtml"/>
         </referenceContainer>
+        <block class="Magento\Framework\View\Element\Template" name="head.critical" as="head.critical" template="Magento_Theme::html/container.phtml"/>
         <block class="Magento\Framework\View\Element\Template" name="head.additional" as="head.additional" template="Magento_Theme::html/container.phtml"/>
     </body>
```
Comment (mrtuvn) [inline app/code/Magento/Catalog/view/frontend/layout/default.xml:70]: in my opinion should add head.critical in scope of module Magento_Theme. And head.additional should be move too. Just small change you can ignore me at this point to avoid unnecessary change

**Comment by in-session** [comment]
@magento give me instance

**Comment by in-session** [comment]
@magento give me 2.4-develop instance

**Comment by mrtuvn** [comment]
From my understand all critical css will need add to file critical.css in css folder in theme right ? Current we don't have any directive/variables for specific critical styles on less file

**Comment by in-session** [comment]
> From my understand all critical css will need add to file critical.css in css folder in theme right ? Current we don't have any directive/variables for specific critical styles on less file Example
> 
> ```css-less
> & when(@critical = true) {
> //css for critical
> }
> 
> & when(@non-critical = true) {
> //css for case non critical
> }
> ```
Yes the critical css must be created by a tool:
https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/css-topics/css-critical-path.html

**Review by mrtuvn — APPROVED** [review]
looks good to me. Approved. Can you cover with some test coverage

**Comment by mrtuvn** [comment]
increase to P2 because in my opinion it's should have more important than P3

**Comment by engcom-November** [comment]
@in-session, can you do the test coverage for the same.

**Comment by engcom-November** [comment]
@magento create issue

**Comment by engcom-November** [comment]
Hello @in-session,

Thanks for the collaboration & contribution!

**✔️ QA Passed**

**Preconditions:**
Enable `use_css_critical_path`.

**Steps to reproduce:**
- Navigate to Home Page
- Open Inspect, look for criticalCss data-type.

**Before: ✖️** 
image

**After: ✔️** 
image

**Comment by engcom-November** [comment]
Since builds are failing moving it to **Extended Testing**.

**Comment by engcom-November** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-November** [comment]
The repeating failures in Functional tests are flaky, and are known issues.

**Functional Tests B2B :**
**1st Run:**
image

**2nd Run:**
image

Known issues:
[ACQE-6523](https://jira.corp.adobe.com/browse/ACQE-6523) - AdminReorderWithCatalogPriceRuleDiscountTest


**Functional Tests CE :**
**1st Run:**
image

**2nd Run:**
image

Known issues:
[ACQE-6590](https://jira.corp.adobe.com/browse/ACQE-6590) - AdminCheckConfigurationForPayPalExpressCheckoutInUnitesStatesTest

**Functional Tests EE :**
**1st Run:**
image

**2nd Run:**
image

Known issues:
[ACQE-6523](https://jira.corp.adobe.com/browse/ACQE-6523) - AdminReorderWithCatalogPriceRuleDiscountTest
[ACQE-6331](https://jira.corp.adobe.com/browse/ACQE-6331) - StorefrontCreateOrderAllQuantityGroupedProductOptionDefaultStockTest

Hence moving this issue to **Merge in Progress**

**Comment by in-session** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by hostep** [comment]
PR isn't marked as merged (probably because of the additional commits after it got approved), but seems to be merged in this bigger merge: https://github.com/magento/magento2/commit/ada8d1d461204d7e1065627384109df1bfa4dcc9

**Comment by engcom-Charlie** [comment]
Hi @hostep,

Thats right, This PR got merge in https://github.com/magento/magento2/commit/ada8d1d461204d7e1065627384109df1bfa4dcc9.

The PR status is Open as few commits are not part of merged PR at the time of acceptance. I will be adding these remaining commits in next Mainline PR till then requesting for not to close it.

Thank you!
